README.md
[![Build](https://github.com/sarmkadan/coolify-cli/actions/workflows/build.yml/badge.svg)](https://github.com/sarmkadan/coolify-cli/actions/workflows/build.yml)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)
[![.NET](https://img.shields.io/badge/.NET-10.0-purple.svg)](https://dotnet.microsoft.com/)

# Coolify CLI

A powerful, production-ready .NET 10 command-line interface for managing Coolify infrastructure directly from the terminal. Deploy applications, manage databases, monitor health, and stream logs with ease.

## Table of Contents

- [Overview](#overview)
- [Features](#features)
- [Architecture](#architecture)
- [Installation](#installation)
- [Quick Start](#quick-start)
- [Configuration](#configuration)
- [Usage Examples](#usage-examples)
- [CLI Reference](#cli-reference)
- [Configuration Reference](#configuration-reference)
- [Advanced Topics](#advanced-topics)
- [Troubleshooting](#troubleshooting)
- [Testing](#testing)
- [Performance](#performance)
- [Related Projects](#related-projects)
- [Contributing](#contributing)
- [License](#license)

## Overview

Coolify CLI is a modern command-line tool for DevOps engineers, system administrators, and developers who manage infrastructure through the Coolify platform. Built with .NET 10 and System.CommandLine, it provides a fast, reliable, and intuitive interface for all major Coolify operations without leaving the terminal.

### Key Use Cases

- **Continuous Deployment**: Automate application deployments in CI/CD pipelines
- **Infrastructure Monitoring**: Check health status and performance metrics in real-time
- **Database Management**: Provision, configure, and maintain databases across environments
- **Log Analysis**: Stream application and database logs for debugging and monitoring
- **Environment Management**: Manage configuration and secrets across deployments
- **Backup Operations**: Automate database backups and recovery procedures

### Why Coolify CLI?

- **Zero Configuration**: Automatic configuration via environment variables
- **Fast**: Optimized for low latency, built with .NET 10's latest performance features
- **Type-Safe**: Leverages C# for compile-time safety and intellisense
- **Cross-Platform**: Runs on Linux, macOS, and Windows
- **Scriptable**: Perfect for automation, cron jobs, and CI/CD pipelines
- **Error Handling**: Comprehensive error recovery and validation
- **Production-Ready**: Used in production environments managing thousands of deployments

## Features

### Application Management
- **List Applications**: View all deployed applications with status and metadata
- **Get Details**: Retrieve comprehensive application configuration and history
- **Deploy**: Trigger deployments with pre-flight checks and rollback support
- **Manage Environment Variables**: Set, update, and remove application configuration
- **View Logs**: Stream real-time logs with filtering and formatting options

### Database Operations
- **List Databases**: Discover all managed database instances
- **Health Monitoring**: Real-time health checks, performance metrics, and connection pooling status
- **Backups**: Schedule, list, and restore database backups
- **Configuration**: Manage database settings, scaling, and maintenance windows
- **Performance Insights**: CPU usage, memory allocation, active connections, error rates

### Advanced Deployment Features
- **Safe Deployments**: Automated pre-deployment health checks
- **Rollback Support**: Automatic rollback on deployment failure
- **Health Verification**: Post-deployment health validation
- **Deployment Strategies**: Support for blue-green, canary, and rolling deployments
- **Resource Quotas**: Monitor and enforce resource limits

### Monitoring & Observability
- **System Health Checks**: Verify API connectivity and system status
- **Performance Metrics**: Track response times, CPU, memory, and error rates
- **Health Dashboards**: Aggregated health status across applications and databases
- **Event Publishing**: Webhook support for deployment events and alerts

## Architecture

### Layered Architecture Diagram

```
┌─────────────────────────────────────────┐
│   Presentation Layer                    │
│   (Program.cs - System.CommandLine)     │
└──────────────┬──────────────────────────┘
               │
┌──────────────┴──────────────────────────┐
│   Command Processing Layer              │
│   (Commands/ - Business Logic)          │
├─ AdvancedAppCommands                   │
├─ DatabaseManagementCommands            │
├─ MonitoringCommands                    │
└──────────────┬──────────────────────────┘
               │
┌──────────────┴──────────────────────────┐
│   Service Layer                         │
│   (Services/ - Orchestration)           │
├─ ApplicationService                    │
├─ DatabaseService                       │
├─ LogService                            │
├─ HealthCheckService                    │
├─ DeploymentOrchestrator                │
├─ EnvironmentVariableService            │
└──────────────┬──────────────────────────┘
               │
┌──────────────┴──────────────────────────┐
│   Middleware Layer                      │
│   (Middleware/)                         │
├─ AuthenticationMiddleware              │
├─ ErrorHandlingMiddleware               │
├─ LoggingMiddleware                     │
├─ RateLimitingMiddleware                │
└──────────────┬──────────────────────────┘
               │
┌──────────────┴──────────────────────────┐
│   Data Access Layer                     │
│   (Data/ - Repository Pattern)          │
├─ ApplicationRepository                 │
├─ DatabaseRepository                    │
└──────────────┬──────────────────────────┘
               │
┌──────────────┴──────────────────────────┐
│   External Layer                        │
│   (Integration/)                        │
├─ CoolifyApiClient (HTTP)               │
├─ WebhookHandler                        │
└─────────────────────────────────────────┘
```

### Design Patterns Used

- **Repository Pattern**: Data access abstraction
- **Service Layer**: Business logic separation
- **Middleware Chain**: Cross-cutting concerns
- **Dependency Injection**: Loose coupling and testability
- **Factory Pattern**: HttpClient creation and management
- **Command Pattern**: CLI command encapsulation
- **Strategy Pattern**: Deployment strategies (blue-green, canary, rolling)

## Installation

### Prerequisites

- **.NET 10 SDK** or later ([download](https://dotnet.microsoft.com/download/dotnet/10.0))
- **Git** for cloning the repository
- **Coolify API Key** from your Coolify instance

### From Source

```bash
# Clone the repository
git clone https://github.com/Sarmkadan/coolify-cli.git
cd coolify-cli

# Build the project
dotnet build

# Publish for release
dotnet publish -c Release -o ./publish

# Optional: Create a symlink or add to PATH
sudo ln -s /path/to/coolify-cli/publish/coolify-cli /usr/local/bin/coolify-cli
```

### Using Package Manager

```bash
# macOS with Homebrew
brew tap Sarmkadan/coolify-cli
brew install coolify-cli

# Linux with snap
snap install coolify-cli
```

### Docker Installation

```bash
# Pull the Docker image
docker pull sarmkadan/coolify-cli:latest

# Run in container
docker run --rm sarmkadan/coolify-cli:latest coolify-cli version
```

### Verify Installation

```bash
coolify-cli version
# Output: Coolify CLI v1.0.0
# Author: Vladyslav Zaiets
# Website: https://sarmkadan.com
```

## Quick Start

```bash
# 1. Set credentials
export COOLIFY_API_KEY="sk_prod_xxxxxxxxxxxxxxxxxx"
export COOLIFY_API_URL="https://coolify.example.com"

# 2. Verify connectivity
coolify-cli health
# ✓ Connected to Coolify API

# 3. List your applications
coolify-cli app list

# 4. Deploy an application
coolify-cli app deploy <ID>

# 5. Tail its logs
coolify-cli logs <ID> --follow
```

See the [Usage Examples](#usage-examples) section for more commands.

## Configuration

### Environment Variables

Configure the CLI via environment variables:

```bash
# Required
export COOLIFY_API_KEY="your-api-key-here"
export COOLIFY_API_URL="https://your-coolify-instance.com"

# Optional
export COOLIFY_VERBOSE="true"              # Enable verbose logging (default: false)
export COOLIFY_TIMEOUT="30"                # Request timeout in seconds (default: 30)
export COOLIFY_CACHE_ENABLED="true"        # Enable response caching (default: true)
export COOLIFY_CACHE_TTL="300"             # Cache TTL in seconds (default: 300)
export COOLIFY_RETRY_ATTEMPTS="3"          # Number of retry attempts (default: 3)
export COOLIFY_RETRY_DELAY_MS="1000"       # Delay between retries in ms (default: 1000)
```

### .env File Support

Create a `.env` file in your working directory or project root:

```bash
COOLIFY_API_KEY=sk_prod_xxxxxxxxxxxxxxxxxx
COOLIFY_API_URL=https://coolify.example.com
COOLIFY_VERBOSE=true
COOLIFY_TIMEOUT=60
```

### Configuration Validation

The CLI validates configuration on startup:

```bash
coolify-cli --verbose
# Validates API key, URL format, and connectivity
```

## Usage Examples

### 1. List All Applications

```bash
coolify-cli app list

# Output:
# ID    Name                      Status       Deployed At
# ──────────────────────────────────────────────────────
# 1     production-api            running      2026-05-04 14:30:22
# 2     staging-web               stopped      2026-05-02 09:15:45
# 3     demo-dashboard            running      2026-05-03 16:45:00
```

### 2. Get Application Details

```bash
coolify-cli app get 1

# Output:
# Application: production-api
# ID: 1
# Repository: git@github.com:company/api.git
# Branch: main
# Status: running
# Environment: prod-eu
# Created: 2026-01-15 10:30:00
# Last Deployed: 2026-05-04 14:30:22
# Ports: 3000, 8080
# Health Check: http://localhost:3000/health
```

### 3. Deploy an Application

```bash
coolify-cli app deploy 1

# Output:
# [INFO] Coolify CLI v1.0.0
# [INFO] Starting deployment of production-api
# [INFO] Deployment initiated successfully
# Deployment ID: deploy_abc123xyz789
```

### 4. Monitor Deployment Status

```bash
coolify-cli app status 1

# Output:
# Application: production-api
# Current Status: deploying
# Progress: 65%
# Started At: 2026-05-04 14:30:22
# Estimated Time: 2 minutes remaining
```

### 5. List Databases

```bash
coolify-cli db list

# Output:
# ID    Name                 Type          Host                  Status
# ────────────────────────────────────────────────────────────────────
# 1     prod-postgres        postgresql    db.prod.example.com   Healthy
# 2     cache-redis          redis         cache.example.com     Healthy
# 3     backup-mysql         mysql         backup.example.com    Unhealthy
```

### 6. Check Database Health

```bash
coolify-cli db health 1

# Output:
# Database Health Check:
# Status: Healthy
# Response Time: 12ms
# CPU Usage: 23.45%
# Memory: 512.34MB
# Active Connections: 42
# Error Rate: 0.12%
```

### 7. Stream Application Logs

```bash
coolify-cli logs 1 --lines 50

# Output shows last 50 log lines with color coding:
# [2026-05-04 14:35:22] [INFO] Application started
# [2026-05-04 14:35:23] [DEBUG] Loading configuration
# [2026-05-04 14:35:25] [INFO] Server listening on port 3000
# [2026-05-04 14:35:45] [WARN] High memory usage detected
```

### 8. Stream Live Logs

```bash
coolify-cli logs 1 --follow

# Continuously streams logs until interrupted with Ctrl+C
```

### 9. Manage Environment Variables

```bash
coolify-cli app env set 1 DATABASE_URL "postgresql://user:pass@host/db"
coolify-cli app env get 1
coolify-cli app env delete 1 DATABASE_URL
```

### 10. System Health Check

```bash
coolify-cli health

# Output:
# ✓ Connected to Coolify API
# ✓ System health check passed
```

## CLI Reference

### Global Options

```bash
coolify-cli [OPTIONS] COMMAND [ARGS]

Options:
  --verbose, -v      Enable verbose logging
  --help             Show help information
  --version          Display version information
```

### Application Commands

```bash
coolify-cli app list                          # List all applications
coolify-cli app get <ID>                      # Get application details
coolify-cli app deploy <ID>                   # Deploy application
coolify-cli app stop <ID>                     # Stop application
coolify-cli app restart <ID>                  # Restart application
coolify-cli app status <ID>                   # Get deployment status
coolify-cli app logs <ID> [--lines N]         # View application logs
coolify-cli app env get <ID>                  # List environment variables
coolify-cli app env set <ID> <KEY> <VALUE>    # Set environment variable
coolify-cli app env delete <ID> <KEY>         # Delete environment variable
```

### Database Commands

```bash
coolify-cli db list                           # List all databases
coolify-cli db get <ID>                       # Get database details
coolify-cli db health <ID>                    # Check database health
coolify-cli db backup list <ID>               # List backups
coolify-cli db backup create <ID>             # Create backup
coolify-cli db backup restore <ID> <BACKUP>   # Restore backup
coolify-cli db logs <ID> [--lines N]          # View database logs
```

### Log Commands

```bash
coolify-cli logs <APP_ID> [--lines N]         # View application logs
coolify-cli logs <APP_ID> --follow            # Stream logs continuously
coolify-cli logs <APP_ID> --filter ERROR      # Filter logs by level
```

### System Commands

```bash
coolify-cli health                            # System health check
coolify-cli version                           # Display version
coolify-cli config                            # Show configuration
```

## Configuration Reference

### Required Configuration

| Variable | Description | Example |
|----------|-------------|---------|
| `COOLIFY_API_KEY` | Authentication token | `sk_prod_xxx` |
| `COOLIFY_API_URL` | API endpoint URL | `https://coolify.example.com` |

### Optional Configuration

| Variable | Default | Range | Description |
|----------|---------|-------|-------------|
| `COOLIFY_VERBOSE` | `false` | `true/false` | Enable verbose logging |
| `COOLIFY_TIMEOUT` | `30` | `1-300` | Request timeout (seconds) |
| `COOLIFY_CACHE_ENABLED` | `true` | `true/false` | Response caching |
| `COOLIFY_CACHE_TTL` | `300` | `0-3600` | Cache TTL (seconds) |
| `COOLIFY_RETRY_ATTEMPTS` | `3` | `0-10` | Retry attempts |
| `COOLIFY_RETRY_DELAY_MS` | `1000` | `100-5000` | Retry delay (ms) |

## Advanced Topics

### Scripting and Automation

Use Coolify CLI in bash scripts for automation:

```bash
#!/bin/bash
# Deploy all applications in a directory

export COOLIFY_API_KEY="sk_prod_xxx"
export COOLIFY_API_URL="https://coolify.example.com"

for app_id in 1 2 3 4 5; do
    echo "Deploying application $app_id..."
    coolify-cli app deploy "$app_id"
    sleep 10
done

echo "All deployments initiated"
```

### CI/CD Integration

Integrate with GitHub Actions, GitLab CI, or Jenkins:

```yaml
# GitHub Actions example
- name: Deploy with Coolify CLI
  env:
    COOLIFY_API_KEY: ${{ secrets.COOLIFY_API_KEY }}
    COOLIFY_API_URL: ${{ secrets.COOLIFY_API_URL }}
  run: |
    coolify-cli app deploy ${{ env.APP_ID }}
```

### Exit Codes

The CLI returns specific exit codes:

- `0`: Success
- `1`: General error
- `2`: Configuration error
- `3`: API communication error
- `4`: Deployment error
- `5`: Validation error

### Output Formatting

Control output format with flags:

```bash
coolify-cli app list --format json        # JSON output
coolify-cli app list --format csv         # CSV output
coolify-cli app list --format table       # Table format (default)
```

### Rate Limiting

API requests are rate-limited. The CLI handles this automatically with exponential backoff. Adjust retry settings if needed:

```bash
export COOLIFY_RETRY_ATTEMPTS="5"
export COOLIFY_RETRY_DELAY_MS="2000"
```

## Troubleshooting

### Connection Issues

**Problem**: "Failed to connect to Coolify API"

**Solution**:
```bash
# Verify API URL is accessible
curl -I https://your-coolify-instance.com

# Check API key validity
coolify-cli health --verbose

# Verify network connectivity
ping your-coolify-instance.com
```

### Authentication Errors

**Problem**: "Invalid or missing API key"

**Solution**:
```bash
# Verify API key is set
echo $COOLIFY_API_KEY

# Check for leading/trailing whitespace
export COOLIFY_API_KEY="sk_prod_xxx"  # Correct format

# Regenerate API key in Coolify dashboard
```

### Timeout Errors

**Problem**: "Request timeout"

**Solution**:
```bash
# Increase timeout value
export COOLIFY_TIMEOUT="60"

# Check network latency
ping -c 5 your-coolify-instance.com

# Verify API server is responsive
curl -w "@curl-format.txt" -o /dev/null https://your-coolify-instance.com
```

### Deployment Failures

**Problem**: "Deployment failed"

**Solution**:
```bash
# Check application logs
coolify-cli logs 1 --lines 100

# Verify application health
coolify-cli app get 1

# Check system health
coolify-cli health

# Review deployment history
coolify-cli app status 1
```

### Memory Issues

**Problem**: CLI consumes excessive memory

**Solution**:
```bash
# Disable response caching for large datasets
export COOLIFY_CACHE_ENABLED="false"

# Paginate results if listing many items
coolify-cli app list --page 1 --limit 50
```

## Project Structure

```
coolify-cli/
├── Models/                      # Domain entities and models
│   ├── ApiResponse.cs
│   ├── ApplicationDeployment.cs
│   ├── DatabaseConfiguration.cs
│   ├── DeploymentContext.cs
│   ├── EnvironmentVariable.cs
│   ├── Enums.cs
│   ├── LogEntry.cs

CHANGELOG.md
# Changelog

## [2.0.2] - 2026-05-21
### Fixed
- Nullable dereference on `Console.ReadLine` in `env sync` confirmation prompt

## [2.0.0] - 2026-05-01
### Added
- Infrastructure-as-code commands: `iac validate`, `iac diff`, `iac apply`, `iac init`
- YAML template engine with `${VAR}` variable substitution
- Deployment diff preview: `app diff`
- Resource usage monitoring: `resources show`, `resources watch`
- Interactive TUI: `coolify-cli tui`
- `env sync` command for bulk environment variable sync from file

### Changed
- Improved error messages with actionable suggestions on all commands
- Color output auto-disabled when stdout is redirected or `NO_COLOR` is set

## [0.6.0] - 2025-06-16
### Added
- Output formatters: table, JSON, CSV (`--format` flag on list commands)
- Collection, DateTime, and Enum extension helpers

## [0.4.0] - 2025-04-07
### Added
- System health check (`coolify-cli health`)
- Database management: `db list`, `db health`, `db backup`, `db logs`

## [0.2.0] - 2025-02-03
### Added
- `DatabaseService` with list, health, backup, and restore operations
- Improved error output on API 4xx/5xx responses

## [0.1.0] - 2025-01-06
### Added
- Initial release: `app list/get/deploy/stop/restart`, `env get/set/delete`, `logs`, `version`
- `CoolifyApiClient`, `ApplicationService`, `EnvironmentVariableService`, `LogService`
- Configuration via `COOLIFY_API_KEY` / `COOLIFY_API_URL` environment variables
- Color-coded console output, `--verbose` flag


Models/IacTemplate.cs
#nullable enable
namespace CoolifyCli.Models;

using YamlDotNet.Serialization;

/// <summary>
/// Root document model for a declarative infrastructure-as-code YAML template.
/// Describes the desired state of applications and databases to reconcile with Coolify.
/// </summary>
public record InfrastructureTemplate
{
    /// <summary>Gets the schema version of the template document.</summary>
    [YamlMember(Alias = "apiVersion")]
    public string ApiVersion { get; init; } = "v2";

    /// <summary>Gets the resource kind discriminator; must be <c>CoolifyInfrastructure</c>.</summary>
    [YamlMember(Alias = "kind")]
    public string Kind { get; init; } = "CoolifyInfrastructure";

    /// <summary>Gets the identification and labelling metadata for the template.</summary>
    [YamlMember(Alias = "metadata")]
    public required IacTemplateMetadata Metadata { get; init; }

    /// <summary>Gets the list of application resources to provision or reconcile.</summary>
    [YamlMember(Alias = "applications")]
    public List<IacTemplateApplication> Applications { get; init; } = [];

    /// <summary>Gets the list of managed database resources to provision or reconcile.</summary>
    [YamlMember(Alias = "databases")]
    public List<IacTemplateDatabase> Databases { get; init; } = [];

    /// <summary>
    /// Validates the template document structure, yielding an error message for every violation.
    /// Returns an empty sequence when the template is structurally valid.
    /// </summary>
    public IEnumerable<string> Validate()
    {
        if (string.IsNullOrWhiteSpace(Metadata?.Name))
            yield return "metadata.name is required";

        if (ApiVersion is not ("v1" or "v2"))
            yield return $"Unsupported apiVersion '{ApiVersion}'. Accepted values: v1, v2";

        if (Kind != "CoolifyInfrastructure")
            yield return $"Invalid kind '{Kind}'. Expected 'CoolifyInfrastructure'";

        if (Applications.Count == 0 && Databases.Count == 0)
            yield return "Template must define at least one application or database";

        var seenApps = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var app in Applications)
        {
            foreach (var error in app.Validate())
                yield return $"applications['{app.Name}']: {error}";

            if (!seenApps.Add(app.Name))
                yield return $"Duplicate application name: '{app.Name}'";
        }

        var seenDbs = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        foreach (var db in Databases)
        {
            foreach (var error in db.Validate())
                yield return $"databases['{db.Name}']: {error}";

            if (!seenDbs.Add(db.Name))
                yield return $"Duplicate database name: '{db.Name}'";
        }
    }
}

/// <summary>
/// Metadata block providing identification, labelling, and environment targeting for a template.
/// </summary>
public record IacTemplateMetadata
{
    /// <summary>Gets the unique name for this infrastructure stack.</summary>
    [YamlMember(Alias = "name")]
    public required string Name { get; init; }

    /// <summary>Gets an optional human-readable description of the template's purpose.</summary>
    [YamlMember(Alias = "description")]
    public string? Description { get; init; }

    /// <summary>Gets the target environment name (e.g., production, staging, development).</summary>
    [YamlMember(Alias = "environment")]
    public string? Environment { get; init; }

    /// <summary>Gets the semantic version of this template document for change tracking.</summary>
    [YamlMember(Alias = "version")]
    public string? Version { get; init; }

    /// <summary>Gets arbitrary key-value pairs for organisational tagging and filtering.</summary>
    [YamlMember(Alias = "labels")]
    public Dictionary<string, string> Labels { get; init; } = [];
}

/// <summary>
/// Declarative specification for a single application resource.
/// Maps to an <see cref="ApplicationDeployment"/> in the Coolify environment.
/// </summary>
public record IacTemplateApplication
{
    /// <summary>Gets the unique application identifier within this template.</summary>
    [YamlMember(Alias = "name")]
    public required string Name { get; init; }

    /// <summary>Gets the source repository URL (HTTPS or SSH) for the application.</summary>
    [YamlMember(Alias = "repository")]
    public required string Repository { get; init; }

    /// <summary>Gets the git branch to build and deploy from.</summary>
    [YamlMember(Alias = "branch")]
    public string Branch { get; init; } = "main";

    /// <summary>Gets the application runtime environment.</summary>
    [YamlMember(Alias = "runtime")]
    public RuntimeEnvironment Runtime { get; init; } = RuntimeEnvironment.Docker;

    /// <summary>
    /// Gets the Coolify environment identifier this application belongs to.
    /// Falls back to the <c>COOLIFY_ENVIRONMENT_ID</c> environment variable when omitted.
    /// </summary>
    [YamlMember(Alias = "environmentId")]
    public string? EnvironmentId { get; init; }

    /// <summary>Gets the build command used to compile or package the application.</summary>
    [YamlMember(Alias = "buildCommand")]
    public string? BuildCommand { get; init; }

    /// <summary>Gets the command used to start the application process.</summary>
    [YamlMember(Alias = "startCommand")]
    public string? StartCommand { get; init; }

    /// <summary>Gets the list of port numbers exposed by the application container.</summary>
    [YamlMember(Alias = "ports")]
    public List<int> Ports { get; init; } = [];

    /// <summary>Gets the HTTP health probe configuration for this application.</summary>
    [YamlMember(Alias = "healthCheck")]
    public IacHealthCheckSpec? HealthCheck { get; init; }

    /// <summary>Gets environment variable key-value pairs to inject at runtime.</summary>
    [YamlMember(Alias = "environment")]
    public Dictionary<string, string> Environment { get; init; } = [];

    /// <summary>Gets the instance count and auto-scaling policy configuration.</summary>
    [YamlMember(Alias = "scaling")]
    public IacScalingSpec? Scaling { get; init; }

    /// <summary>Gets CPU and memory resource limit constraints for the container.</summary>
    [YamlMember(Alias = "resources")]
    public IacResourceLimits? Resources { get; init; }

    /// <summary>
    /// Validates the application specification, yielding an error string for every violation.
    /// </summary>
    internal IEnumerable<string> Validate()
    {
        if (string.IsNullOrWhiteSpace(Name))
            yield return "name is required";

        if (string.IsNullOrWhiteSpace(Repository))
            yield return "repository is required";
        else if (!Uri.TryCreate(Repository, UriKind.Absolute, out _))
            yield return $"repository '{Repository}' is not a valid URL";

        foreach (var port in Ports.Where(p => p is < 1 or > 65535))
            yield return $"port {port} is out of the valid range 1–65535";

        if (Scaling is { Instances: < 1 })
            yield return "scaling.instances must be at least 1";

        if (HealthCheck is { IntervalSeconds: < 5 or > 3600 })
            yield return "healthCheck.intervalSeconds must be between 5 and 3600";
    }
}

/// <summary>
/// HTTP probe configuration for application liveness and readiness checking.
/// </summary>
public record IacHealthCheckSpec
{
    /// <summary>Gets the HTTP path to probe for health (e.g., <c>/health</c>).</summary>
    [YamlMember(Alias = "url")]
    public required string Url { get; init; }

    /// <summary>Gets the number of seconds between consecutive health probe attempts.</summary>
    [YamlMember(Alias = "intervalSeconds")]
    public int IntervalSeconds { get; init; } = 30;

    /// <summary>
    /// Gets the number of consecutive probe failures required to declare the application unhealthy.
    /// </summary>
    [YamlMember(Alias = "failureThreshold")]
    public int FailureThreshold { get; init; } = 3;
}

/// <summary>
/// Instance scaling specification controlling the desired replica count and auto-scaling policy.
/// </summary>
public record IacScalingSpec
{
    /// <summary>Gets the desired number of running application instances.</summary>
    [YamlMember(Alias = "instances")]
    public int Instances { get; init; } = 1;

    /// <summary>Gets the policy that governs automatic scale-out and scale-in behaviour.</summary>
    [YamlMember(Alias = "policy")]
    public ScalingPolicy Policy { get; init; } = ScalingPolicy.Manual;
}

/// <summary>
/// CPU and memory resource limits applied to an application container.
/// Values follow Kubernetes-style notation (e.g., <c>500m</c>, <c>512Mi</c>, <c>2Gi</c>).
/// </summary>
public record IacResourceLimits
{
    /// <summary>
    /// Gets the CPU limit in millicores notation (e.g., <c>250m</c> = 0.25 vCPU,
    /// <c>1000m</c> = 1 vCPU).
    /// </summary>
    [YamlMember(Alias = "cpuLimit")]
    public string? CpuLimit { get; init; }

    /// <summary>Gets the memory limit using SI binary suffixes (e.g., <c>512Mi</c>, <c>2Gi</c>).</summary>
    [YamlMember(Alias = "memoryLimit")]
    public string? MemoryLimit { get; init; }
}

/// <summary>
/// Declarative specification for a managed database resource.
/// Maps to a <see cref="DatabaseConfiguration"/> in the Coolify environment.
/// </summary>
public record IacTemplateDatabase
{
    /// <summary>Gets the unique database identifier within this template.</summary>
    [YamlMember(Alias = "name")]
    public required string Name { get; init; }

    /// <summary>Gets the database management system type to provision.</summary>
    [YamlMember(Alias = "type")]
    public DatabaseType Type { get; init; }

    /// <summary>Gets the engine version to use (e.g., <c>"15"</c> for PostgreSQL 15).</summary>
    [YamlMember(Alias = "version")]
    public string? Version { get; init; }

    /// <summary>Gets the maximum number of concurrent database connections to allow.</summary>
    [YamlMember(Alias = "maxConnections")]
    public int? MaxConnections { get; init; }

    /// <summary>Gets the connection timeout in seconds for establishing new connections.</summary>
    [YamlMember(Alias = "connectionTimeoutSeconds")]
    public int? ConnectionTimeoutSeconds { get; init; }

    /// <summary>Gets the automated backup configuration for this database.</summary>
    [YamlMember(Alias = "backup")]
    public IacBackupSpec? Backup { get; init; }

    /// <summary>
    /// Validates the database specification, yielding an error string for every violation.
    /// </summary>
    internal IEnumerable<string> Validate()
    {
        if (string.IsNullOrWhiteSpace(Name))
            yield return "name is required";

        if (MaxConnections.HasValue && MaxConnections.Value is < 1 or > 1000)
            yield return "maxConnections must be between 1 and 1000";

        if (ConnectionTimeoutSeconds.HasValue && ConnectionTimeoutSeconds.Value is < 5 or > 300)
            yield return "connectionTimeoutSeconds must be between 5 and 300";

        if (Backup is { RetentionDays: < 1 or > 365 })
            yield return "backup.retentionDays must be between 1 and 365";
    }
}

/// <summary>
/// Automated backup configuration for a database, specifying strategy, cron schedule, and retention.
/// </summary>
public record IacBackupSpec
{
    /// <summary>Gets whether automated backups are active for this database.</summary>
    [YamlMember(Alias = "enabled")]
    public bool Enabled { get; init; } = false;

    /// <summary>Gets the backup strategy to apply.</summary>
    [YamlMember(Alias = "strategy")]
    public BackupStrategy Strategy { get; init; } = BackupStrategy.Snapshot;

    /// <summary>Gets the number of days to retain backup archives before automatic deletion.</summary>
    [YamlMember(Alias = "retentionDays")]
    public int RetentionDays { get; init; } = 30;

    /// <summary>
    /// Gets the cron expression that controls when backups run (e.g., <c>"0 2 * * *"</c> = 02:00 daily).
    /// Defaults to the Coolify platform default when omitted.
    /// </summary>
    [YamlMember(Alias = "schedule")]
    public string? Schedule { get; init; }
}

// ─── Result types ─────────────────────────────────────────────────────────────

/// <summary>
/// Outcome of structurally validating an <see cref="InfrastructureTemplate"/>, accumulating
/// both hard errors (which block application) and advisory warnings.
/// </summary>
public record TemplateValidationResult
{
    /// <summary>Gets whether the template passed all validation rules with no errors.</summary>
    public bool IsValid => Errors.Count == 0;

    /// <summary>Gets the collected validation error messages that must be resolved before applying.</summary>
    public List<string> Errors { get; init; } = [];

    /// <summary>Gets advisory warning messages that do not block application but indicate risk.</summary>
    public List<string> Warnings { get; init; } = [];

    /// <summary>Gets the template name as read from <c>metadata.name</c>, if available.</summary>
    public string? TemplateName { get; init; }
}

/// <summary>
/// Result of comparing an <see cref="InfrastructureTemplate"/> against the live Coolify state,
/// categorised into additions, modifications, removals, and unchanged resources.
/// </summary>
public record TemplateDiffResult
{
    /// <summary>Gets resources declared in the template that do not yet exist in the live environment.</summary>
    public List<TemplateDiffEntry> Added { get; init; } = [];

    /// <summary>Gets resources that exist in both the template and live environment but differ.</summary>
    public List<TemplateDiffEntry> Modified { get; init; } = [];

    /// <summary>
    /// Gets resources present in the live environment but absent from the template.
    /// These are flagged for awareness but are never automatically removed.
    /// </summary>
    public List<TemplateDiffEntry> Removed { get; init; } = [];

    /// <summary>Gets resources that are identical between the template and live state.</summary>
    public List<TemplateDiffEntry> Unchanged { get; init; } = [];

    /// <summary>Gets whether any additions or modifications require reconciliation.</summary>
    public bool HasChanges => Added.Count > 0 || Modified.Count > 0;
}

/// <summary>
/// Represents a single resource entry within a <see cref="TemplateDiffResult"/>.
/// </summary>
public record TemplateDiffEntry
{
    /// <summary>Gets the resource category, such as <c>Application</c> or <c>Database</c>.</summary>
    public required string ResourceType { get; init; }

    /// <summary>Gets the name of the resource.</summary>
    public required string Name { get; init; }

    /// <summary>Gets a human-readable description of the detected change, if applicable.</summary>
    public string? ChangeDescription { get; init; }
}

/// <summary>
/// Aggregated outcome of applying an <see cref="InfrastructureTemplate"/>, summarising
/// every resource-level operation attempted during the reconciliation phase.
/// </summary>
public record TemplateApplyResult
{
    /// <summary>Gets the ordered list of operations that were attempted during apply.</summary>
    public List<TemplateApplyOperation> Operations { get; init; } = [];

    /// <summary>Gets whether all attempted operations completed successfully.</summary>
    public bool Success => Operations.Count > 0 && Operations.All(o => o.Succeeded);

    /// <summary>Gets the total elapsed time from validation through to the final operation.</summary>
    public TimeSpan Duration { get; init; }

    /// <summary>Gets the number of operations that succeeded.</summary>
    public int SucceededCount => Operations.Count(o => o.Succeeded);

    /// <summary>Gets the number of operations that failed.</summary>
    public int FailedCount => Operations.Count(o => !o.Succeeded);
}

/// <summary>
/// Describes the outcome of a single resource-level create or update operation performed
/// during a template apply phase.
/// </summary>
public record TemplateApplyOperation
{
    /// <summary>Gets the resource category that was operated on.</summary>
    public required string ResourceType { get; init; }

    /// <summary>Gets the name of the specific resource that was operated on.</summary>
    public required string ResourceName { get; init; }

    /// <summary>
    /// Gets the action label for this operation (e.g., <c>Create</c>, <c>Update</c>).
    /// </summary>
    public required string Action { get; init; }

    /// <summary>Gets whether this individual operation completed without error.</summary>
    public bool Succeeded { get; init; }

    /// <summary>Gets an optional message providing detail about the outcome.</summary>
    public string? Message { get; init; }
}
